platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/bls/protocol/CrsProtocol.java

```diff
@@ -0,0 +1,239 @@
+        this.protocolManager.addRound(this::verifyAndOpen);
+        this.protocolManager.setDisqualificationCleanup(this::cleanupAfterDisqualification);
+        this.protocolManager.setViabilityChecker(this::isProtocolViable);
+        this.protocolManager.setFinishingMethod(this::performFinish);
+
+        this.commitments = new HashMap<>();
+        this.bilinearMap = bilinearMap;
+
+        this.randomGroupElements = new RandomGroupElements(bilinearMap, protocolManager.getRandom());
+    }
+
+    /**
+     * <strong>ROUND 1</strong>
+     * <p>
+     * Broadcasts a commitment to the random group elements
+     *
+     * @return the {@link CommitmentMessage}
+     */
+    private ProtocolMessage broadcastCommitment(final List<ProtocolMessage> inputMessages) {
+        try {
+            return new CommitmentMessage(nodeId, randomGroupElements.commit());
+        } catch (final NoSuchAlgorithmException e) {
+            protocolManager.errorOccurred();
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * <strong>ROUND 2</strong>
+     * <p>
+     * Verifies that an honest majority of parties broadcast a {@link CommitmentMessage} in round 1. If this is
+     * determined to be true, openly reveal random group elements
+     *
+     * @param inputMessages a list of {@link ProtocolMessage}s received in round 1
+     * @return an {@link OpeningMessage}
+     */
+    private ProtocolMessage verifyAndOpen(final List<ProtocolMessage> inputMessages) {
+        final List<CommitmentMessage> commitmentMessages = protocolManager.filterCast(
+                inputMessages, addressBook.getSortedNodeIds(), CommitmentMessage.class, true);
+
+        // Gather the commitments from the messages
+        commitmentMessages.forEach(message -> commitments.put(message.getSenderId(), message.getCommitment()));
+
+        // Open my random elements
+        return new OpeningMessage(nodeId, randomGroupElements);
+    }
+
+    /**
+     * <p>Throws an error if not enough commitments were received in the last round</p>
+     */
+    private Crs performFinish(final List<ProtocolMessage> inputMessages) {
+        // Cast messages to correct type
+        final List<OpeningMessage> openingMessages =
+                protocolManager.filterCast(inputMessages, addressBook.getSortedNodeIds(), OpeningMessage.class, true);
+
+        final List<GroupElement> generator1Elements = new ArrayList<>();
+        final List<GroupElement> generator2Elements = new ArrayList<>();
+
+        for (final OpeningMessage message : openingMessages) {
+            final NodeId dealer = message.getSenderId();
+
+            final byte[] commitment;
+            try {
+                // Recompute commitment from the opening
+                commitment = message.getRandomGroupElements().commit();
+            } catch (final NoSuchAlgorithmException e) {
+                protocolManager.errorOccurred();
+                throw new RuntimeException(e);
+            }
```

Comment:
I've never seen a `NoSuchAlgorithmException` be recoverable in any our use cases. This is the type of exception I think the java team messed up on by us required to catch it. I'd suggest squelching this to a standard runtime exception deeper in the API.

Reply:
fine with that

Reply:
I am now passing a digest object into the protocol, so the exception is handled when that's constructed